imports/wasi_snapshot_preview1/poll.go

```diff
@@ -72,88 +81,148 @@ func pollOneoffFn(ctx context.Context, mod api.Module, params []uint64) syscall.
 	_ /* precision */ = le.Uint64(inBuf[16:24]) // Unused
 	flags := le.Uint16(inBuf[24:32])
 
+	var err syscall.Errno
 	// subclockflags has only one flag defined:  subscription_clock_abstime
 	switch flags {
 	case 0: // relative time
 	case 1: // subscription_clock_abstime
-		return syscall.ENOTSUP
+		err = syscall.ENOTSUP
 	default: // subclockflags has only one flag defined.
-		return syscall.EINVAL
+		err = syscall.EINVAL
 	}
 
-	// https://linux.die.net/man/3/clock_settime says relative timers are
-	// unaffected. Since this function only supports relative timeout, we can
-	// skip name ID validation and use a single sleep function.
+	if err != 0 {
+		return 0, err
+	} else {
+		// https://linux.die.net/man/3/clock_settime says relative timers are
+		// unaffected. Since this function only supports relative timeout, we can
+		// skip name ID validation and use a single sleep function.
 
-	sysCtx := mod.(*wasm.ModuleInstance).Sys
-	sysCtx.Nanosleep(int64(timeout))
-	return 0
+		return time.Duration(timeout), 0
+	}
 }
 
-// processFDEvent returns a validation error or syscall.ENOTSUP as file or socket
-// subscriptions are not yet supported.
-func processFDEvent(mod api.Module, eventType byte, inBuf []byte) syscall.Errno {
-	fd := le.Uint32(inBuf)
+func processFDEvent(mod api.Module, argBuf []byte, v *rwSub) bool {
 	fsc := mod.(*wasm.ModuleInstance).Sys.FS()
+	fd := le.Uint32(argBuf)
 
 	// Choose the best error, which falls back to unsupported, until we support
 	// files.
-	errno := syscall.ENOTSUP
-	if eventType == wasip1.EventTypeFdRead {
-		if f, ok := fsc.LookupFile(fd); ok {
+	if v.eventType == wasip1.EventTypeFdRead {
+		if f, ok := fsc.LookupFile(v.fd); ok {
 			st, _ := f.Stat()
 			// if fd is a pipe, then it is not a char device (a tty)
-			if st.Mode&fs.ModeCharDevice != 0 {
-				errno = syscall.EBADF
+			if fd == 0 && st.Mode&fs.ModeCharDevice == 0 {
+				v.errno = byte(wasip1.ErrnoSuccess)
+			} else {
+				// is a tty
+				return true
 			}
 		} else {
-			errno = syscall.EBADF
+			v.errno = byte(wasip1.ErrnoBadf)
 		}
-	} else if eventType == wasip1.EventTypeFdWrite && internalsys.WriterForFile(fsc, fd) == nil {
-		errno = syscall.EBADF
+	} else if v.eventType == wasip1.EventTypeFdWrite && internalsys.WriterForFile(fsc, v.fd) == nil {
+		v.errno = byte(wasip1.ErrnoBadf)
 	}
+	return false
+}
 
-	return errno
+func write(outBuf []byte, value *rwSub) {
+	// Write the event corresponding to the processed subscription.
+	// https://github.com/WebAssembly/WASI/blob/snapshot-01/phases/snapshot/docs.md#-event-struct
+	copy(outBuf, value.userData)            // userdata
+	outBuf[value.outOffset+8] = value.errno // uint16, but safe as < 255
+	outBuf[value.outOffset+9] = 0
+	le.PutUint32(outBuf[value.outOffset+10:], uint32(value.eventType))
+	// TODO: When FD events are supported, write outOffset+16
+}
+
+// validateFDEvent returns a validation error or syscall.ENOTSUP as file or socket
```

Comment:
in practice we are now returning ENOTSUP in fewer cases than we originally did -- we really are supporting (some) FD events in some way...